src/baldr/shortcut_recovery.h

```diff
@@ -0,0 +1,231 @@
+    for (const auto& level : valhalla::baldr::TileHierarchy::levels()) {
+      // we dont get shortcuts on level 2 and up
+      if (level.level > 1)
+        continue;
+      // for each tile
+      for (auto tile_id : reader->GetTileSet(level.level)) {
+        // cull cache if we are over allocated
+        if (reader->OverCommitted())
+          reader->Trim();
+        // for each edge in the tile
+        const auto* tile = reader->GetGraphTile(tile_id);
+        for (const auto& edge : tile->GetDirectedEdges()) {
+          // skip non-shortcuts or the shortcut is one we wont use
+          if (!edge.shortcut())
+            continue;
+          auto shortcut_id = tile->header()->graphid();
+          shortcut_id.set_id(&edge - tile->directededge(0));
+          // skip already found opposing edges
+          if (shortcuts.find(shortcut_id) != shortcuts.end())
+            continue;
+          // recover the shortcut and make a copy for opposing direction
+          auto recovered = recover_shortcut(*reader, shortcut_id);
+          decltype(recovered) opp_recovered = recovered;
+          std::reverse_copy(recovered.cbegin(), recovered.cend(), opp_recovered.begin());
+          // save some stats
+          bool failed = recovered.front() == shortcut_id;
+          unrecovered += failed;
+          superseded += failed ? 0 : recovered.size();
+          // cache it even if it failed (no point in trying the same thing twice)
+          shortcuts.emplace(shortcut_id, std::move(recovered));
+
+          // its cheaper to get the opposing without crawling the graph
+          const auto* opp_tile = tile;
+          auto opp_id = reader->GetOpposingEdgeId(shortcut_id, opp_tile);
+          if (!opp_id.Is_Valid())
+            continue; // dont store edges which arent in our tileset
+
+          for (auto& id : opp_recovered) {
+            id = reader->GetOpposingEdgeId(id, opp_tile);
+            if (!id.Is_Valid()) {
+              opp_recovered = {opp_id};
+              break;
+            }
+          }
+          // stats
+          failed = opp_recovered.front() == opp_id;
+          unrecovered += failed;
+          superseded += failed ? 0 : opp_recovered.size();
+          // cache it even if it failed (no point in trying the same thing twice)
+          shortcuts.emplace(opp_id, std::move(opp_recovered));
+        }
+      }
+    }
+
+    LOG_INFO(std::to_string(shortcuts.size()) + " shortcuts recovered as " +
+             std::to_string(superseded) + " superseded edges. " + std::to_string(unrecovered) +
+             " shortcuts could not be recovered.");
+  }
+
+  /**
+   * Recovers the edges comprising a shortcut edge.
+   * @param reader       The graphreader for graph data ccess
+   * @param  shortcutid  Graph Id of the shortcut edge.
+   * @return Returns the edgeids of the directed edges this shortcut represents.
+   */
+  std::vector<valhalla::baldr::GraphId>
+  recover_shortcut(valhalla::baldr::GraphReader& reader,
```

Comment:
this method was moved from graphreader into here, so that the cache can call it to fill the cache or call it on the fly in the event cache is disabled (the default)